src/commons/api/import.rs

```diff
@@ -0,0 +1,147 @@
+//! Data types used to support importing a CA structure for testing or automated set ups.
+
+use std::collections::HashMap;
+
+use serde::{Deserialize, Deserializer};
+
+use rpki::{
+    ca::idexchange::{CaHandle, ParentHandle},
+    repository::resources::ResourceSet,
+    uri,
+};
+
+use crate::{
+    commons::api::PublicationServerUris,
+    daemon::{ca::ta_handle, config},
+};
+
+use super::RoaConfiguration;
+
+/// This type contains the full structure of CAs and signed objects etc that is
+/// set up when the import API is used.
+#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
+pub struct Structure {
+    pub ta_aia: uri::Rsync,
+    pub ta_uri: uri::Https,
+    pub publication_server_uris: PublicationServerUris,
+    pub cas: Vec<ImportCa>,
+}
+
+impl Structure {
+    pub fn new(
+        ta_aia: uri::Rsync,
+        ta_uri: uri::Https,
+        publication_server_uris: PublicationServerUris,
+        cas: Vec<ImportCa>,
+    ) -> Self {
+        Structure {
+            ta_aia,
+            ta_uri,
+            publication_server_uris,
+            cas,
+        }
+    }
+    // Check that all parents are valid for the CAs in this structure
+    // in the order in which they appear, and that the parent CAs have
+    // the resources for each child CA.
+    pub fn validate_ca_hierarchy(&self) -> Result<(), String> {
+        let mut seen: HashMap<ParentHandle, ResourceSet> = HashMap::new();
+        // ta is implied
+        seen.insert(ta_handle().into_converted(), ResourceSet::all());
+        for ca in &self.cas {
+            let mut ca_resources = ResourceSet::empty();
+            for parent in &ca.parents {
+                if let Some(seen_parent_resources) = seen.get(parent.handle()) {
+                    if seen_parent_resources.contains(&parent.resources) {
```

Comment:
Shouldn't this check if the parent contains the resources claimed by the CA? It seems to check that the parent contains the parents resources which should always be true...

Reply:
I can see how this was confusing.

That "parent" is the parent as the CA sees it. But, I added some comments and renamed "parent" to "ca_parent". Hopefully that makes it more clear what is going on.